### PR TITLE
xen-guest-agent: init at 0.3.0

### PR DESCRIPTION
**Pull request by matdibu**
## Description of changes

New upstream project aimed at replacing xe-guest-utilities.

Homepage: https://gitlab.com/xen-project/xen-guest-agent

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/virtualization/xen-guest-agent/default.nix [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  src = fetchFromGitLab {
+    owner = "xen-project";
+    repo = pname;
+    rev = "e6e165e74ce161704af7c96966a26f62a93bdbed";
+    hash = "sha256-MUcYtrLJdTkcgBtjHv1wcOJQxK0xNb9SItEo6yBWJMU=";
+  };
+
+  cargoHash = "sha256-pov7+mmlpu/PnbtY40yWh9KskbcP3SDpVailarARY6E=";
+
+  LIBCLANG_PATH="${llvmPackages.libclang.lib}/lib";
```
Comment (SuperSandro2000) [inline pkgs/tools/virtualization/xen-guest-agent/default.nix:21]: Env variables have, like the other things, usually a space before and after the equals sign and should ideally be under the env attr.

**Review by SuperSandro2000 — APPROVED** [review]

File: pkgs/tools/virtualization/xen-guest-agent/default.nix (hunk 1 above)
Comment (matdibu) [reply pkgs/tools/virtualization/xen-guest-agent/default.nix]: fixed, thank you!

**Comment by SuperSandro2000** [comment]
Sorry that I noticed so late but the review commits should be squashed into the init commit.
Please see the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes) for exact details.

Other than that I think this can be merged.

**Comment by matdibu** [comment]
> Sorry that I noticed so late but the review commits should be squashed into the init commit. Please see the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes) for exact details.
> 
> Other than that I think this can be merged.

Done, and PGP signed.

**Review by SuperSandro2000 — APPROVED** [review]

**Review by pbsds — COMMENTED** [review]
Seems to build and run properly, but I haven't tested in a xen VM

```
[nix-shell:/dev/shm/nixpkgs-review/pr-272966]$ ldd result/bin/xen-guest-agent  
	linux-vdso.so.1 (0x00007ffcc548a000)
	libgcc_s.so.1 => /nix/store/dghjv6hfz0s0z4kffa5ahyw2mhp79215-gcc-12.3.0-lib/lib/libgcc_s.so.1 (0x00007fb25ff8c000)
	libm.so.6 => /nix/store/9y8pmvk8gdwwznmkzxa6pwyah52xy3nk-glibc-2.38-27/lib/libm.so.6 (0x00007fb25feac000)
	libc.so.6 => /nix/store/9y8pmvk8gdwwznmkzxa6pwyah52xy3nk-glibc-2.38-27/lib/libc.so.6 (0x00007fb25f818000)
	/nix/store/9y8pmvk8gdwwznmkzxa6pwyah52xy3nk-glibc-2.38-27/lib/ld-linux-x86-64.so.2 => /nix/store/9y8pmvk8gdwwznmkzxa6pwyah52xy3nk-glibc-2.38-27/lib64/ld-linux-x86-64.so.2 (0x00007fb25ffaf000)

[nix-shell:/dev/shm/nixpkgs-review/pr-272966]$ RUST_BACKTRACE=full result/bin/xen-guest-agent  
Error: Os { code: 2, kind: NotFound, message: "No such file or directory" }

[nix-shell:/dev/shm/nixpkgs-review/pr-272966]$ RUST_BACKTRACE=full strace result/bin/xen-guest-agent  |& grep ENOENT | tail -n2
access("/dev/xen/xenbus", F_OK)         = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/proc/xen/xenbus", 0x7ffcd7e015e0, 0) = -1 ENOENT (No such file or directory)
```


It seems sandro's request was misunderstood. 
Please split the pull request into two commits, titled:

* `maintainers: add matdibu`
* `xen-guest-agent: init at 0.3.0-dev`



---


- [x] package path fits guidelines
- [x] package name fits guidelines
- [x] package version fits guidelines
- [x] package builds on x86_64-linux
- [x] executables tested on x86_64-linux
- [x] `meta.description` is set and fits guidelines
- [x] `meta.license` fits upstream license
- [x] `meta.platforms` is set
- [x] `meta.maintainers` is set
- [x] build time only dependencies are declared in `nativeBuildInputs`
- [x] source is fetched using the appropriate function
- [x] the list of `phases` is not overridden
- [x] when a phase (like `installPhase`) is overridden it starts with `runHook preInstall` and ends with `runHook postInstall`.
- [x] patches have a comment describing either the upstream URL or a reason why the patch wasn't upstreamed
- [x] patches that are remotely available are fetched rather than vendored

**Comment by matdibu** [comment]
@pbsds I have tested it locally, on a machine running `xcp-ng 8.3 beta 1`, within a NixOS `24.05.20231216.7d69298` guest.

In Xen Orchestra it shows the agent string as `Management agent 1.0.0-proto-0.3.0 detected`, and it properly reports everything:
- OS name
- OS kernel
- CPU/Memory/Network/Disk stats
- IP address

Which is an improvement over the old `xe-guest-utilities` agent which did not report the IP address and and lists itself as `Management agent version	@PRODUCT_MAJOR_VERSION@.@PRODUCT_MINOR_VERSION@.@PRODUCT_MICRO_VERSION@-@NUMERIC_BUILD_NUMBER@`

This is the log of `sudo ./result/bin/xen-guest-agent --stderr --loglevel debug`:
```
[2023-12-17T15:54:36Z INFO  xenstore_rs::libxenstore] Loading libxenstore.so.3.0
[2023-12-17T15:54:36Z DEBUG os_info::imp::lsb_release] lsb_release command failed with Os { code: 2, kind: NotFound, message: "No such file or directory" }
[2023-12-17T15:54:36Z DEBUG netlink_proto::handle] handle: forwarding new request to connection
[2023-12-17T15:54:36Z DEBUG netlink_proto::connection] reading incoming messages
[2023-12-17T15:54:36Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-12-17T15:54:36Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-12-17T15:54:36Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-12-17T15:54:36Z DEBUG netlink_proto::connection] handling requests
[2023-12-17T15:54:36Z DEBUG netlink_proto::connection] sending messages
[2023-12-17T15:54:36Z DEBUG netlink_proto::connection] reading incoming messages
[2023-12-17T15:54:36Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-12-17T15:54:36Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] handling messages (request id = RequestId { sequence_number: 1, port: 0 })
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] handling response to request RequestId { sequence_number: 1, port: 0 }
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] done handling response to request RequestId { sequence_number: 1, port: 0 }
[2023-12-17T15:54:36Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] handling messages (request id = RequestId { sequence_number: 1, port: 0 })
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] handling response to request RequestId { sequence_number: 1, port: 0 }
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] done handling response to request RequestId { sequence_number: 1, port: 0 }
[2023-12-17T15:54:36Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-12-17T15:54:36Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] handling messages (request id = RequestId { sequence_number: 1, port: 0 })
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] handling response to request RequestId { sequence_number: 1, port: 0 }
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] done handling response to request RequestId { sequence_number: 1, port: 0 }
[2023-12-17T15:54:36Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-12-17T15:54:36Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-12-17T15:54:36Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-12-17T15:54:36Z DEBUG netlink_proto::connection] handling requests
[2023-12-17T15:54:36Z DEBUG netlink_proto::connection] sending messages
[2023-12-17T15:54:36Z DEBUG xen_guest_agent::vif_detect] reading /sys/class/net/lo/device/devtype: No such file or directory (os error 2)
[2023-12-17T15:54:36Z DEBUG xen_guest_agent::collector_net] NewLink(RefCell { value: NetInterface { index: 1, name: "lo", toolstack_iface: None } } Some("00:00:00:00:00:00"))
[2023-12-17T15:54:36Z DEBUG xen_guest_agent::collector_net] NewLink(RefCell { value: NetInterface { index: 2, name: "enX0", toolstack_iface: Vif(0) } } Some("ee:7d:f4:9b:22:f1"))
[2023-12-17T15:54:36Z DEBUG netlink_proto::handle] handle: forwarding new request to connection
[2023-12-17T15:54:36Z DEBUG netlink_proto::connection] reading incoming messages
[2023-12-17T15:54:36Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-12-17T15:54:36Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-12-17T15:54:36Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-12-17T15:54:36Z DEBUG netlink_proto::connection] handling requests
[2023-12-17T15:54:36Z DEBUG netlink_proto::connection] sending messages
[2023-12-17T15:54:36Z DEBUG netlink_proto::connection] reading incoming messages
[2023-12-17T15:54:36Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-12-17T15:54:36Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] handling messages (request id = RequestId { sequence_number: 2, port: 0 })
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] handling response to request RequestId { sequence_number: 2, port: 0 }
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] done handling response to request RequestId { sequence_number: 2, port: 0 }
[2023-12-17T15:54:36Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] handling messages (request id = RequestId { sequence_number: 2, port: 0 })
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] handling response to request RequestId { sequence_number: 2, port: 0 }
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] done handling response to request RequestId { sequence_number: 2, port: 0 }
[2023-12-17T15:54:36Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] handling messages (request id = RequestId { sequence_number: 2, port: 0 })
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] handling response to request RequestId { sequence_number: 2, port: 0 }
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] done handling response to request RequestId { sequence_number: 2, port: 0 }
[2023-12-17T15:54:36Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-12-17T15:54:36Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] handling messages (request id = RequestId { sequence_number: 2, port: 0 })
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] handling response to request RequestId { sequence_number: 2, port: 0 }
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] done handling response to request RequestId { sequence_number: 2, port: 0 }
[2023-12-17T15:54:36Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] handling messages (request id = RequestId { sequence_number: 2, port: 0 })
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] handling response to request RequestId { sequence_number: 2, port: 0 }
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] done handling response to request RequestId { sequence_number: 2, port: 0 }
[2023-12-17T15:54:36Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] handling messages (request id = RequestId { sequence_number: 2, port: 0 })
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] handling response to request RequestId { sequence_number: 2, port: 0 }
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] done handling response to request RequestId { sequence_number: 2, port: 0 }
[2023-12-17T15:54:36Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-12-17T15:54:36Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] handling messages (request id = RequestId { sequence_number: 2, port: 0 })
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] handling response to request RequestId { sequence_number: 2, port: 0 }
[2023-12-17T15:54:36Z DEBUG netlink_proto::protocol::protocol] done handling response to request RequestId { sequence_number: 2, port: 0 }
[2023-12-17T15:54:36Z DEBUG netlink_proto::codecs] NetlinkCodec: decoding next message
[2023-12-17T15:54:36Z DEBUG netlink_proto::connection] forwarding unsolicited messages to the connection handle
[2023-12-17T15:54:36Z DEBUG netlink_proto::connection] forwaring responses to previous requests to the connection handle
[2023-12-17T15:54:36Z DEBUG netlink_proto::connection] handling requests
[2023-12-17T15:54:36Z DEBUG netlink_proto::connection] sending messages
[2023-12-17T15:54:36Z DEBUG xen_guest_agent::collector_net] NewAddress(RefCell { value: NetInterface { index: 1, name: "lo", toolstack_iface: None } } 127.0.0.1)
[2023-12-17T15:54:36Z DEBUG xen_guest_agent::collector_net] NewAddress(RefCell { value: NetInterface { index: 2, name: "enX0", toolstack_iface: Vif(0) } } 192.168.0.154)
[2023-12-17T15:54:36Z DEBUG xen_guest_agent::collector_net] NewAddress(RefCell { value: NetInterface { index: 2, name: "enX0", toolstack_iface: Vif(0) } } 192.168.0.161)
[2023-12-17T15:54:36Z DEBUG xen_guest_agent::collector_net] NewAddress(RefCell { value: NetInterface { index: 1, name: "lo", toolstack_iface: None } } ::1)
[2023-12-17T15:54:36Z DEBUG xen_guest_agent::collector_net] NewAddress(RefCell { value: NetInterface { index: 2, name: "enX0", toolstack_iface: Vif(0) } } fe80::ec7d:f4ff:fe9b:22f1)
[2023-12-17T15:54:36Z DEBUG xen_guest_agent::collector_net] NewAddress(RefCell { value: NetInterface { index: 2, name: "enX0", toolstack_iface: Vif(0) } } fe80::66e4:ee26:23b3:16b5)
[2023-12-17T15:54:36Z DEBUG xen_guest_agent::xenstore_schema_std] AddMac not applied
```

**Comment by pbsds** [comment]
Result of `nixpkgs-review pr 272966` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>xen-guest-agent</li>
  </ul>
</details>

**Pull request merged**